src/methods/correlation/chatterjee.jl

```diff
@@ -0,0 +1,212 @@
+correlation coefficient.
+
+## Estimation
+
+- [Example 1](@ref example_ChatterjeeCorrelation). Estimating the Chatterjee correlation coefficient
+    for independent and for dependent variables. 
+- [Example 2](@ref example_SurrogateAssociationTest_ChatterjeeCorrelation). Testing the significance
+    of a Chatterjee-type association using a surrogate test.
+"""
+struct ChatterjeeCorrelation{XS, R, L, DR, YR, YSI, YSS, RNG} <: CorrelationMeasure
+    # Pre-allocated containers (if input data are given).
+    # If no input data are given, then these are all `nothing`.
+    xs_inds::XS # indices that would sort `x`
+    r::R # ranks
+    𝓁::L # 𝓁s
+    Δr::DR # absolute value of first differences of ranks
+    y_sortedbyx::YR # ys, sorted according to the order that sorts the xs.
+    y_sortedbyx_inds::YSI # A container holding the indices that would sort the *second* input vector (Y).
+    y_sortedbyx_sorted::YSS # the sorted version of `y_sortedbyx`
+    rng::RNG
+    handle_ties::Bool
+    lt::Function
+end
+
+# Non-preallocated version
+function ChatterjeeCorrelation(; handle_ties = true, rng = Random.default_rng(), lt = (a, b) -> isless_rand(rng, a, b))
+    return ChatterjeeCorrelation{Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(rng)}(
+        nothing, nothing, nothing, nothing, nothing, nothing, nothing, rng, handle_ties, lt)
+end
+
+# Preallocated version
+function ChatterjeeCorrelation(x, y; 
+        handle_ties = true, rng = Random.default_rng(), lt = (a, b) -> isless_rand(rng, a, b))
+    verify_inputs_chatterjee(x, y)
+    n = length(x)
+    xs_inds = zeros(Int, n)
+    y_sortedbyx = similar(y)  # rearranged `y` (sorted according to `xs_inds`)
+    y_sortedbyx_inds = zeros(Int, n)  # rearranged `y` (sorted according to `xs_inds`)
+    y_sortedbyx_sorted = similar(y) # sort the rearranged `y`.
+
+    𝓁 = zeros(Int, n)
+    r = zeros(Int, n)
+    Δr = zeros(n)
+    return ChatterjeeCorrelation(xs_inds, r, 𝓁, Δr, y_sortedbyx, y_sortedbyx_inds, 
+        y_sortedbyx_sorted, rng, handle_ties, lt)
+end
+
+# If the user provided a non-preallocated version, then we simply create a new version
+# of the measure that *is* pre-allocated and use that.
+function association(m::ChatterjeeCorrelation{Nothing}, x, y)
+    measure = ChatterjeeCorrelation(x, y, rng = m.rng, lt = m.lt, handle_ties = m.handle_ties)
+    return association(measure, x, y)
+end
+
+# Pre-allocated version
+function association(m::ChatterjeeCorrelation{<:AbstractVector}, x, y)
+    verify_inputs_chatterjee(x, y)
+    n = length(x)
+
+    # Rearrange input data according to paper.
+    m.xs_inds[:] = sortperm!(m.xs_inds, x, lt = m.lt) 
+    m.y_sortedbyx[:] = y[m.xs_inds]
+
+    # Sort the rearranged `y`s so that we can more efficiently find the ranks.
+    m.y_sortedbyx_inds[:] = sortperm!(m.y_sortedbyx_inds, m.y_sortedbyx)
+    m.y_sortedbyx_sorted[:] = m.y_sortedbyx[m.y_sortedbyx_inds]
```

Comment:
why do you do `x[:] = y` instead of `x .= y`?

Reply:
Good question. I updated the PR using `.=` notation.